### PR TITLE
Feature/33491 cost range merge

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/33491

This builds the base used to solve the ticket issue, review this first.

### CODE REVIEW

File: src/Tribe/Cost_Utils.php [hunk 1]
```diff
@@ -338,4 +338,78 @@ public function parse_cost_range( $costs, $max_decimals = null ) {
...
+		$merging_cost               = (array) $merging_cost;
+		$numeric_merging_cost_costs = array();
+		foreach ( $merging_cost as $c ) {
+			if ( is_numeric( $c ) ) {
+				$numeric_merging_cost_costs[] = $c;
+			}
+		}
+
+		$matches = array();
+		preg_match_all( '!\d+(?:\.\d+)?!', $original_string_cost, $matches );
```
Comment (barryhughes) [inline src/Tribe/Cost_Utils.php]: Given `$10.99 - $59.95` looks like we'd get two matches back - `10.99` and `59.95` - but if the original string cost was `€10,99 - €59,95` we'd get four matches - `10`, `99`, `59` and `95`. Is that a concern or is that handled upstream/elsewhere?

File: src/Tribe/Cost_Utils.php [hunk 2]
```diff
@@ -338,4 +338,78 @@ public function parse_cost_range( $costs, $max_decimals = null ) {
...
+	 * @param array              $sorted_maxs          An array of non numeric price maximums sorted smaller to larger,
+	 *                                                 e.g. `['Donation min $10', 'Donation min $20', 'Donation min
+	 *                                                 $100']`.
+	 *
+	 * @return string|array The merged cost range.
+	 */
+	public function merge_cost_ranges( $original_string_cost, $merging_cost, $with_currency_symbol, $sorted_mins = array(), $sorted_maxs = array() ) {
+		if ( empty( $merging_cost ) || $original_string_cost === $merging_cost ) {
+			return $original_string_cost;
+
```
Comment (barryhughes) [inline src/Tribe/Cost_Utils.php]: Extra line we can remove!

File: src/Tribe/Cost_Utils.php [hunk 3]
```diff
@@ -338,4 +338,78 @@ public function parse_cost_range( $costs, $max_decimals = null ) {
 
 		return (array) $ocost;
 	}
+
+	/**
+	 * @param       string       $original_string_cost A string cost with or without currency symbol,
+	 *                                                 e.g. `10 - 20`, `Free` or `2$ - 4$`.
+	 * @param       array|string $merging_cost         A single string cost representation to merge or a an array of
```
Comment (barryhughes) [inline src/Tribe/Cost_Utils.php]: Typo: _or a an array_ should probably be _or an array_

File: tests/functional/Tribe/Events/Cost_UtilsTest.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	public function test_merge_cost_ranges_into( $expected, $original_string_cost = '', $merging_cost = '', $with_currency_symbol = false, $sorted_mins = array(), $sorted_maxs = array() ) {
+
+		add_filter( 'tribe_currency_symbol',
+			function () {
+				return '$';
+			},
+			100 );
+		add_filter( 'tribe_reverse_currency_position', '__return_false');
+
+		$sut = $this->make_instance();
```
Comment (barryhughes) [inline tests/functional/Tribe/Events/Cost_UtilsTest.php]: Curious - why `$sut` as the var name here?

File: tests/functional/Tribe/Events/Cost_UtilsTest.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	 * @test
+	 * it should be instantiatable
+	 */
+	public function it_should_be_instantiatable() {
+		$sut = $this->make_instance();
+
+		$this->assertInstanceOf( 'Tribe__Events__Cost_Utils', $sut );
+	}
+
+	public function merge_cost_ranges_into_ranges() {
```
Comment (barryhughes) [inline tests/functional/Tribe/Events/Cost_UtilsTest.php]: Is it worth covering values with a decimal point (and different formats of decimal points) and should we cover use of thousands separators?

File: tests/functional/Tribe/Events/Cost_UtilsTest.php (hunk 4 above)
Comment (lucatume) [inline tests/functional/Tribe/Events/Cost_UtilsTest.php]: Subject Under Test

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Comma required after last value in array declaration | [Cost_Utils.php:394](https://github.com/moderntribe/the-events-calendar/blob/beb7c84d3218f09688b8497efc834a5e80005085/src/Tribe/Cost_Utils.php#L394)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
I've added tests and handling code for the cases you pointed out.
I will take 15' to look into the thousands separator and see what can be made of it.
All good points yours Barry, thanks.

**Comment by lucatume** [comment]
Ok, there is no easy way I can think of to handle the intricacies of the thousands separator.
A way can be found but not in time budget; I've tried the [`number_format_i18n` function](https://codex.wordpress.org/Function_Reference/number_format_i18n) but that requires numbers to be numbers already.  
Sadly on PHP 5.2 the [NumberFormatter](http://php.net/manual/en/class.numberformatter.php) class is not available.
I'm open to any idea.

**Comment by barryhughes** [comment]
Hmm, not sure.

Maybe as a minimum we document the fact (that we don't handle a mixture of thousands separators and decimals) either as a remark in one of the docblocks or as a failing test that we skip as something to possibly address at some future point?

Totally trust your judgement.

**Comment by lucatume** [comment]
I've added some skipping tests, this is more of a memo for us than something users can know but not control.

**Comment by barryhughes** [comment]
Perfect, I think that leaves us in a reasonable spot :+1:

**Pull request merged**